plonky2/src/gates/selectors.rs

```diff
@@ -0,0 +1,82 @@
+use plonky2_field::extension_field::Extendable;
+use plonky2_field::polynomial::PolynomialValues;
+
+use crate::gates::gate::{GateInstance, GateRef};
+use crate::hash::hash_types::RichField;
+
+/// Placeholder value to indicate that a gate doesn't use a selector polynomial.
+pub(crate) const UNUSED_SELECTOR: usize = u32::MAX as usize;
+
+#[derive(Debug, Clone)]
+pub(crate) struct SelectorsInfo {
+    pub(crate) selector_indices: Vec<usize>,
+    pub(crate) groups: Vec<(usize, usize)>,
+    pub(crate) num_selectors: usize,
+}
+
+/// Returns the selector polynomials and related information.
+///
+/// Selector polynomials are computed as follows:
+/// Partition the gates into (the smallest amount of) groups `{ G_i }`, such that for each group `G`
+/// `|G| + max_{g in G} g.degree() <= max_degree`. These groups are constructed greedily from
+/// the list of gates sorted by degree.
+/// We build a selector polynomial `S_i` for each group `G_i`, with
+/// S_i[j] =
+///     if j-th row gate=g_k in G_i
+///         k
+///     else
+///         UNUSED_SELECTOR
+pub(crate) fn selector_polynomials<F: RichField + Extendable<D>, const D: usize>(
+    gates: &[GateRef<F, D>],
+    instances: &[GateInstance<F, D>],
+    max_degree: usize,
+) -> (Vec<PolynomialValues<F>>, SelectorsInfo) {
+    let n = instances.len();
+
+    // Greedily construct the groups.
+    let mut groups = Vec::new();
+    let mut pos = 0;
```

Comment:
Couple nits,
- I think the loop structure may be slightly more clear if we change `pos`, `i` to `start`, `size` or similar
- Should we panic if `i` never increments because some gate is too high-degree? It seems like the current code might infinite loop in that case, unless we check for it elsewhere (I forget if we do)

Reply:
Good point, I added a check.